src/OrchardCore/OrchardCore.BackgroundTasks/BackgroundHostedService.cs

```diff
@@ -0,0 +1,237 @@
+                                        "Error while processing background task \"{0}\" on tenant \"{1}\".",
+                                        tenant, taskName);
+                                }
+                            }
+                        }
+                    }
+                });
+
+                startUtc = DateTime.UtcNow;
+                await Task.Delay(MinIdleTime, stoppingToken);
+                await pollingDelay;
+            }
+        }
+
+        Task IShellDescriptorManagerEventHandler.Changed(ShellDescriptor descriptor, string tenant)
+        {
+            CleanTenantSchedulers(tenant);
+            return Task.CompletedTask;
+        }
+
+        private IEnumerable<ShellContext> GetRunningShells()
+        {
+            return _shellHost.ListShellContexts()?.Where(s => s.Settings?.State == TenantState.Running)
+                .OrderBy(s => s.Settings?.Name).ToArray() ?? Enumerable.Empty<ShellContext>();
+        }
+
+        private void CleanTenantSchedulers(string tenant)
+        {
+            var keys = _schedulers.Where(kv => kv.Value.Tenant == tenant).Select(kv => kv.Key).ToArray();
+
+            foreach (var key in keys)
+            {
+                _schedulers.TryRemove(key, out var scheduler);
+            }
+        }
+
+        private class Scheduler
+        {
+            public Scheduler(string tenant, IBackgroundTask task, DateTime startUtc)
+            {
+                Tenant = tenant;
+                var attribute = task.GetType().GetCustomAttribute<BackgroundTaskAttribute>();
+                Schedule = attribute?.Schedule ?? "* * * * *";
+                StartUtc = startUtc;
+            }
+
+            public string Tenant { get; }
+            public string Schedule { get; }
+            public DateTime StartUtc { get; private set; }
+
+            public bool ShouldRun()
+            {
+                var now = DateTime.UtcNow;
+
+                if (now >= CrontabSchedule.Parse(Schedule).GetNextOccurrence(StartUtc))
+                {
+                    StartUtc = now;
+                    return true;
+                }
+
+                return false;
+            }
+        }
+    }
+
+    internal static class EnumerableExtensions
+    {
+
+        public static Task ForEachAsync<T>(this IEnumerable<T> source, Func<T, Task> body)
+        {
+            return Task.WhenAll(
+                from partition in Partitioner.Create(source).GetPartitions(8)
+                select Task.Run(async delegate
+                {
+                    using (partition)
+                    {
+                        while (partition.MoveNext())
+                        {
+                            await body(partition.Current);
+                        }
+                    }
+                }));
+        }
+    }
+
+    internal static class ShellExtensions
+    {
+        public static HttpContext GetBackgroundHttpContext(this ShellContext shell)
```

Comment:
It's really the perfect occasion to get rid of the HTTP context in the background tasks stuff :sweat_smile:

Reply:
So, is it ok? I checked, when entering in the hosted service, that httpContext is null, then each shell task will have its own fake httpContext.

Reply:
Actually, I was secretly hoping you'd remove this fake context... :sweat_smile:

Reply:
I would like to please you ;) I'm open to any idea. Hmm, i remember that in O1 we use a workContext which reference an httpContext and i made a specific work context accessor which checks if we are in the background and, as i remember, this is at this point that we return a workContext with a fake httpContext.

Reply:
Ultimately, it's your call. But that would be great to get rid of this pattern in OC. It's too much error-prone: tasks can't have access to a real HTTP context and at best you give them a fake one with very limited data (well, just the host and the path base).

/cc @sebastienros

Reply:
Yes i understand and anyway i will do as you want, knowing that right now there are some services used in the background which use the `RequestServices` to resolve scoped services.

Otherwise, about the fake httpContext it's just a beginning. As i remember, in O1 the most used property in the background was the path (related to a given tenant), and we have a wrapper to override many other props but just because they would be null otherwise.

But in aspnetcore most of the props of the default httpcontext have a non null default value, i will check this more in depth.

Also, in O1 some services just check if they are executed in the background and then just do nothing.